lib/sudo-system/src/time.rs

```diff
@@ -0,0 +1,192 @@
+use std::{
+    io::{Read, Write},
+    mem::MaybeUninit,
+    ops::{Add, Sub},
+};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SystemTime {
+    secs: i64,
+    nsecs: i64,
+}
+
+impl SystemTime {
+    pub(crate) fn new(secs: i64, nsecs: i64) -> SystemTime {
+        let mut secs = secs + nsecs / 1_000_000_000;
+        let mut nsecs = nsecs % 1_000_000_000;
+
+        if secs > 0 && nsecs < 0 {
+            secs -= 1;
+            nsecs += 1_000_000_000;
+        } else if secs < 0 && nsecs > 0 {
+            secs += 1;
+            nsecs -= 1_000_000_000;
+        }
```

Comment:
Why not simply disallow negative values for `nsecs` and go for:
```suggestion
        let secs = secs + nsecs.div_euclid(1_000_000_000);
        let nsecs = nsecs.rem_euclid(1_000_000_000);
```

Reply:
(In practice, lines 18..23 feel like they are going to be dead code.)

Reply:
Update, from the manpage:

`EINVAL (clock_settime()): tp.tv_sec is negative or tp.tv_nsec is outside the range [0..999,999,999]`.

I.e., for minimum surprise the type invariant here should match that (which the `_euclid` functions ensures), instead of going to the 'both arguments should agree in sign'.